core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java

```diff
@@ -0,0 +1,150 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.registry.loader;
+
+import org.geysermc.geyser.GeyserImpl;
+import org.geysermc.geyser.api.event.lifecycle.GeyserLoadResourcePacksEvent;
+import org.geysermc.geyser.api.pack.ResourcePack;
+import org.geysermc.geyser.pack.GeyserResourcePack;
+import org.geysermc.geyser.pack.GeyserResourcePackManifest;
+import org.geysermc.geyser.pack.path.GeyserPathPackCodec;
+import org.geysermc.geyser.text.GeyserLocale;
+import org.geysermc.geyser.util.FileUtils;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/**
+ * Loads {@link ResourcePack}s within a {@link Path} directory, firing the {@link GeyserLoadResourcePacksEvent}.
+ */
+public class ResourcePackLoader implements RegistryLoader<Path, Map<String, ResourcePack>> {
+
+    static final PathMatcher PACK_MATCHER = FileSystems.getDefault().getPathMatcher("glob:**.{zip,mcpack}");
+
+    private static final boolean SHOW_RESOURCE_PACK_LENGTH_WARNING = Boolean.parseBoolean(System.getProperty("Geyser.ShowResourcePackLengthWarning", "true"));
+
+    /**
+     * Loop through the packs directory and locate valid resource pack files
+     */
+    @Override
+    public Map<String, ResourcePack> load(Path directory) {
+        Map<String, ResourcePack> packMap = new HashMap<>();
+
+        if (!Files.exists(directory)) {
+            try {
+                Files.createDirectory(directory);
+            } catch (IOException e) {
+                GeyserImpl.getInstance().getLogger().error("Could not create packs directory", e);
+            }
+
+            return packMap; // As we just created the directory it will be empty
+        }
+
+        List<Path> resourcePacks;
+        try (Stream<Path> stream = Files.walk(directory)) {
+            resourcePacks = stream.filter(PACK_MATCHER::matches)
+                    .collect(Collectors.toCollection(ArrayList::new)); // toList() does not guarantee mutability
```

Comment:
We don't need mutability right? And otherwise `.collect(Collectors.toList())` should be mutable as well.

Reply:
[Collectors#toList - There are no guarantees on the type, mutability, serializability, or thread-safety of the List returned](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#toList--)

Needs to be mutable because its passed to GeyserLoadResourcePacksEvent